src/cpu_treeshap_model.h

```diff
@@ -0,0 +1,420 @@
+
+template <typename tl_threshold_t, typename tl_output_t>
+double
+leaf_probability(
+    const treelite::Tree<tl_threshold_t, tl_output_t>& tree, int node)
+{
+  if (tree.HasSumHess(node) && tree.HasSumHess(0)) {
+    return static_cast<double>(tree.SumHess(node) / tree.SumHess(0));
+  } else if (tree.HasDataCount(node) && tree.HasDataCount(0)) {
+    return static_cast<double>(tree.DataCount(node)) / tree.DataCount(0);
+  }
+  return 0.0f;
+}
+
+// The linear treeshap algorithm requires some extra info for each tree
+template <typename tl_threshold_t, typename tl_output_t>
+struct TreeMetaInfo {
+  std::vector<double> weights;
+  std::vector<int> edge_heights;
+  std::vector<int> parents;
+  int max_depth = 0;
+  const treelite::Tree<tl_threshold_t, tl_output_t>& tree;
+  double global_bias;
+  int class_idx;
+
+  std::pair<int, double> Recurse(
+      int node, int parent = -1, int depth = 0,
+      std::map<int, int> seen_features = std::map<int, int>())
+  {
+    if (node != 0) {
+      auto feature = tree.SplitIndex(parent);
+      double weight = 0.0;
+      if (tree.HasSumHess(parent) && tree.HasSumHess(node)) {
+        weight = static_cast<double>(tree.SumHess(node) / tree.SumHess(parent));
+      } else if (tree.HasDataCount(parent) && tree.HasDataCount(node)) {
+        weight =
+            static_cast<double>(tree.DataCount(node)) / tree.DataCount(parent);
+      } else {
+        throw rapids::TritonException(
+            rapids::Error::Unsupported, "Model does not have node statistics.");
+      }
+      if (seen_features.count(feature)) {
+        auto seen_node_id = seen_features[feature];
+        parents[node] = seen_node_id;
+        weight *= weights[seen_node_id];
+      }
+      weights[node] = weight;
+      seen_features[feature] = node;
+    }
+
+    double bias = 0.0;
+    if (!tree.IsLeaf(node)) {
+      auto [left_max_features, left_bias] =
+          Recurse(tree.LeftChild(node), node, depth + 1, seen_features);
+      auto [right_max_features, right_bias] =
+          Recurse(tree.RightChild(node), node, depth + 1, seen_features);
+      edge_heights[node] = std::max(left_max_features, right_max_features);
+      bias = left_bias + right_bias;
+    } else {
+      edge_heights[node] = seen_features.size();
+      max_depth = std::max(max_depth, depth);
+      double leaf_value = tree.HasLeafVector(node)
+                              ? tree.LeafVector(node)[class_idx]
+                              : tree.LeafValue(node);
+      bias = leaf_probability(tree, node) * leaf_value;
+    }
+    return std::make_pair(edge_heights[node], bias);
+  }
+
+  explicit TreeMetaInfo(
+      const treelite::Tree<tl_threshold_t, tl_output_t>& tree, int class_idx)
+      : tree(tree), class_idx(class_idx)
+  {
+    weights.resize(tree.num_nodes, 1.0f);
+    parents.resize(tree.num_nodes, -1);
+    edge_heights.resize(tree.num_nodes);
+    auto [_, bias] = Recurse(0);
+    global_bias = bias;
+    if (max_depth >= 32) {
```

Comment:
Could we add this prominently to the docs?

Reply:
Actually, if you could generally update the docs mentioning this feature (as experimental) and its limitations, that would be great.